### PR TITLE
inject: fix --default-inbound-policy not setting annotation

### LINKED ISSUE #9168: `--default-inbound-policy` flag ignored by `linkerd inject`
### What is the issue?

The `linkerd inject` CLI command accepts a `--default-inbound-policy` flag, which is supposed to configure the `config.linkerd.io/default-inbound-policy` annotation for the injected resource(s). However, this flag does nothing, and the annotation is not added to the injected resources.

In addition, the `inject` command does not validate the provided policy the way the `install` command does. If an invalid default policy is provided, nothing happens.

### How can it be reproduced?

Injecting emojivoto with `--default-inbound-policy`:
```shell
curl --proto '=https' --tlsv1.2 -sSfL https://run.linkerd.io/emojivoto.yml | linkerd inject --default-inbound-policy deny -
```

Setting an invalid value for `--default-inbound-policy`:
```shell
curl --proto '=https' --tlsv1.2 -sSfL https://run.linkerd.io/emojivoto.yml | linkerd inject --default-inbound-policy asdf -
```

### Logs, error output, etc

Output from the `inject` command above. Note that no `config.linkerd.io/default-inbound-policy` annotations are added:
```yaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    linkerd.io/inject: enabled
  name: emojivoto
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: emoji
  namespace: emojivoto
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: voting
  namespace: emojivoto
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: web
  namespace: emojivoto
---
apiVersion: v1
kind: Service
metadata:
  name: emoji-svc
  namespace: emojivoto
spec:
  ports:
  - name: grpc
    port: 8080
    targetPort: 8080
  - name: prom
    port: 8801
    targetPort: 8801
  selector:
    app: emoji-svc
---
apiVersion: v1
kind: Service
metadata:
  name: voting-svc
  namespace: emojivoto
spec:
  ports:
  - name: grpc
    port: 8080
    targetPort: 8080
  - name: prom
    port: 8801
    targetPort: 8801
  selector:
    app: voting-svc
---
apiVersion: v1
kind: Service
metadata:
  name: web-svc
  namespace: emojivoto
spec:
  ports:
  - name: http
    port: 80
    targetPort: 8080
  selector:
    app: web-svc
  type: ClusterIP
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/name: emoji
    app.kubernetes.io/part-of: emojivoto
    app.kubernetes.io/version: v11
  name: emoji
  namespace: emojivoto
spec:
  replicas: 1
  selector:
    matchLabels:
      app: emoji-svc
      version: v11
  template:
    metadata:
      annotations:
        linkerd.io/inject: enabled
      labels:
        app: emoji-svc
        version: v11
    spec:
      containers:
      - env:
        - name: GRPC_PORT
          value: "8080"
        - name: PROM_PORT
          value: "8801"
        image: docker.l5d.io/buoyantio/emojivoto-emoji-svc:v11
        name: emoji-svc
        ports:
        - containerPort: 8080
          name: grpc
        - containerPort: 8801
          name: prom
        resources:
          requests:
            cpu: 100m
      serviceAccountName: emoji
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/name: vote-bot
    app.kubernetes.io/part-of: emojivoto
    app.kubernetes.io/version: v11
  name: vote-bot
  namespace: emojivoto
spec:
  replicas: 1
  selector:
    matchLabels:
      app: vote-bot
      version: v11
  template:
    metadata:
      annotations:
        linkerd.io/inject: enabled
      labels:
        app: vote-bot
        version: v11
    spec:
      containers:
      - command:
        - emojivoto-vote-bot
        env:
        - name: WEB_HOST
          value: web-svc.emojivoto:80
        image: docker.l5d.io/buoyantio/emojivoto-web:v11
        name: vote-bot
        resources:
          requests:
            cpu: 10m
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/name: voting
    app.kubernetes.io/part-of: emojivoto
    app.kubernetes.io/version: v11
  name: voting
  namespace: emojivoto
spec:
  replicas: 1
  selector:
    matchLabels:
      app: voting-svc
      version: v11
  template:
    metadata:
      annotations:
        linkerd.io/inject: enabled
      labels:
        app: voting-svc
        version: v11
    spec:
      containers:
      - env:
        - name: GRPC_PORT
          value: "8080"
        - name: PROM_PORT
          value: "8801"
        image: docker.l5d.io/buoyantio/emojivoto-voting-svc:v11
        name: voting-svc
        ports:
        - containerPort: 8080
          name: grpc
        - containerPort: 8801
          name: prom
        resources:
          requests:
            cpu: 100m
      serviceAccountName: voting
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/name: web
    app.kubernetes.io/part-of: emojivoto
    app.kubernetes.io/version: v11
  name: web
  namespace: emojivoto
spec:
  replicas: 1
  selector:
    matchLabels:
      app: web-svc
      version: v11
  template:
    metadata:
      annotations:
        linkerd.io/inject: enabled
      labels:
        app: web-svc
        version: v11
    spec:
      containers:
      - env:
        - name: WEB_PORT
          value: "8080"
        - name: EMOJISVC_HOST
          value: emoji-svc.emojivoto:8080
        - name: VOTINGSVC_HOST
          value: voting-svc.emojivoto:8080
        - name: INDEX_BUNDLE
          value: dist/index_bundle.js
        image: docker.l5d.io/buoyantio/emojivoto-web:v11
        name: web-svc
        ports:
        - containerPort: 8080
          name: http
        resources:
          requests:
            cpu: 100m
      serviceAccountName: web
---
```

### output of `linkerd check -o short`

```console
$ just linkerd check -o short
Status check results are √
```

### Environment

```console
$ linkerd version
Client version: edge-22.8.2
Server version: edge-22.8.2

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.9", GitCommit:"c1de2d70269039fe55efb98e737d9a29f9155246", GitTreeState:"archive", BuildDate:"1980-01-01T00:00:00Z", GoVersion:"go1.18.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.7+k3s1", GitCommit:"ac70570999c566ac3507d2cc17369bb0629c1cc0", GitTreeState:"clean", BuildDate:"2021-11-29T16:40:13Z", GoVersion:"go1.16.10", Compiler:"gc", Platform:"linux/amd64"}
WARNING: version difference between client (1.23) and server (1.21) exceeds the supported minor version skew of +/-1

$ k3d version
k3d version v5.4.4
k3s version v1.21.7-k3s1 (default)

$ uname -a
Linux noctis 5.18.12 #1-NixOS SMP PREEMPT_DYNAMIC Fri Jul 15 08:12:06 UTC 2022 x86_64 GNU/Linux
```

### Possible solution

_No response_

### Additional context

_No response_

### Would you like to work on fixing this bug?

_No response_

### PR DESCRIPTION
**Pull request by hawkw**
Depends on #9195

Currently, `linkerd inject --default-inbound-policy` does not set the
`config.linkerd.io/default-inbound-policy` annotation on the injected
resource(s).

The `inject` command does _try_ to set that annotation if it's set in
the `Values` generated by `proxyFlagSet`:
https://github.com/linkerd/linkerd2/blob/14d1dbb3b71b3e2a205b8bed084f13eb79e3d00d/cli/cmd/inject.go#L485-L487

...but, the flag in the proxy `FlagSet` doesn't set
`Values.Proxy.DefaultInboundPolicy`, it sets
`Values.PolicyController.DefaultAllowPolicy`:
https://github.com/linkerd/linkerd2/blob/7c5e3aaf406ec528c68396cc67d67474598c6a84/cli/cmd/options.go#L375-L379

This is because the flag set is shared across `linkerd inject` and
`linkerd install` subcommands, and in `linkerd install`, we want to set
the default policy for the whole cluster by configuring the policy
controller. In `linkerd inject`, though, we want to add the annotation
to the injected pods only.

This branch fixes this issue by changing the flag so that it sets the
`Values.Proxy.DefaultInboundPolicy` instead of the
`Values.PolicyController.DefaultAllowPolicy` value. In `linkerd
install`, we then set `Values.PolicyController.DefaultAllowPolicy` based
on the value of `Values.Proxy.DefaultInboundPolicy`, while in `inject`,
we will now actually add the annotation.

This branch is based on PR #9195, which adds validation to reject
invalid values for `--default-inbound-policy`, rather than on `main`.
This is because the validation code added in that PR had to be moved
around a bit, since it now needs to validate the
`Values.Proxy.DefaultInboundPolicy` value rather than the
`Values.PolicyController.DefaultAllowPolicy` value. I thought using
#9195 as a base branch was better than basing this on `main` and then
having to resolve merge conflicts later. When that PR merges, this can 
be rebased onto `main`.

Fixes #9168

### CODE REVIEW

File: cli/cmd/install.go [hunk 1]
```diff
@@ -198,6 +198,11 @@ func installControlPlane(ctx context.Context, k8sAPI *k8s.KubernetesAPI, w io.Wr
 		return err
 	}
 
+	// the `--default-inbound-policy` flag configures the proxy override
+	// annotation, but in `install`, we also want to configure the default for
+	// the policy controller.
+	values.PolicyController.DefaultAllowPolicy = values.Proxy.DefaultInboundPolicy
```
Comment (adleong) [inline cli/cmd/install.go:204]: I'd recommend getting rid of the values.PolicyController.DefaultAllowPolicy value completely and instead use the values.Proxy.DefaultInboundPolicy value in the policy-controller template.  I don't think it makes sense for these two values to ever have different values.  Fewer values makes it easier for helm users too.

File: cli/cmd/options.go [hunk 2]
```diff
@@ -624,6 +623,12 @@ func validateProxyValues(values *l5dcharts.Values) error {
 		}
 	}
 
+	if values.Proxy.DefaultInboundPolicy != "" {
+		if err := validatePolicy(values.Proxy.DefaultInboundPolicy); err != nil {
+			return err
+		}
+	}
```
Comment (kleimkuhler) [inline cli/cmd/options.go:630]: We'll always want to validate `values.Proxy.DefaultInboundPolicy`. With the `values.Proxy.DefaultInboundPolicy != ""` check, we're allowing users to do `linkerd install --set proxy.DefaultInboundPolicy=""` which isn't right.

```suggestion
if err := validatePolicy(values.Proxy.DefaultInboundPolicy); err != nil {
	return err
}
```

Comment (hawkw) [reply cli/cmd/options.go]: thanks, good catch --- i think the outer `if` was left over from when we still allowed setting both values.

**Review by kleimkuhler — APPROVED** [review]

**Review by adleong — APPROVED** [review]
Tested and it works great.  :+1: subject to nits above

File: charts/partials/templates/_proxy.tpl [hunk 3]
```diff
@@ -38,7 +38,7 @@ env:
 - name: LINKERD2_PROXY_POLICY_WORKLOAD
   value: "$(_pod_ns):$(_pod_name)"
 - name: LINKERD2_PROXY_INBOUND_DEFAULT_POLICY
-  value: {{.Values.proxy.defaultInboundPolicy | default .Values.policyController.defaultAllowPolicy}}
+  value: {{.Values.proxy.defaultInboundPolicy | quote}}
```
[the remark this answers is not in the document]
Comment (hawkw) [reply charts/partials/templates/_proxy.tpl]: i removed this but the golden files are still quoted...

**Review by adleong — APPROVED** [review]

**Pull request merged**